libs/wallet/src/connectors/snap-connector.ts

```diff
@@ -1,247 +1,282 @@
-        return true;
-      }
+      const res = await this.invokeSnap<{ chainID: string }>(
+        JsonRpcMethod.GetChainId,
+        {
+          networkEndpoints: [this.node],
+        }
+      );
+      return { chainId: res.chainID };
     } catch (err) {
-      return false;
+      this.stopPoll();
+      throw ConnectorErrors.chainId;
     }
+  }
 
-    return false;
+  async listKeys() {
+    try {
+      const res = await this.invokeSnap<{
+        keys: Array<{ publicKey: string; name: string }>;
+      }>(JsonRpcMethod.ListKeys);
+      return res.keys;
+    } catch (err) {
+      this.stopPoll();
+      throw ConnectorErrors.listKeys;
+    }
   }
 
-  async sendTx(pubKey: string, transaction: Transaction) {
-    if (!this.nodeAddress) throw SnapConnectorErrors.NODE_ADDRESS_NOT_SET;
-    if (!this.snapId) throw SnapConnectorErrors.SNAP_ID_NOT_SET;
+  async isConnected() {
+    try {
+      // Check if metamask is unlocked
+      if (!window.ethereum.selectedAddress) {
+        throw ConnectorErrors.noWallet;
+      }
 
-    // This step is needed to strip the transaction object from any additional
-    // properties, such as `__proto__`, etc.
-    let txData = null;
+      // If this throws its likely the snap is disabled or has been uninstalled
+      await this.listKeys();
+      return { connected: true };
+    } catch (err) {
+      this.stopPoll();
+      return { connected: false };
+    }
+  }
+
+  async sendTransaction(params: TransactionParams) {
     try {
-      txData = JSON.parse(JSON.stringify(transaction));
+      const res = await this.invokeSnap<{
+        transactionHash: string;
+        transaction: { signature: { value: string } };
+        receivedAt: string;
+        sentAt: string;
+      }>(JsonRpcMethod.SendTransaction, {
+        publicKey: params.publicKey,
+        sendingMode: params.sendingMode,
+        transaction: params.transaction,
+        networkEndpoints: [this.node],
+      });
+
+      return {
+        transactionHash: res.transactionHash,
+        signature: res.transaction.signature.value,
+        receivedAt: res.receivedAt,
+        sentAt: res.sentAt,
+      };
     } catch (err) {
-      throw SnapConnectorErrors.TRANSACTION_PARSE;
+      if (err instanceof ConnectorError) {
+        throw err;
+      }
+
+      throw ConnectorErrors.sendTransaction;
     }
+  }
 
-    const payload = {
-      method: 'client.send_transaction',
-      params: {
-        sendingMode: 'TYPE_SYNC',
-        transaction: txData,
-        publicKey: pubKey,
-        networkEndpoints: [this.nodeAddress],
-      },
-    };
+  on(event: VegaWalletEvent, callback: () => void) {
+    this.pollListeners[event].push(callback);
+  }
 
-    const result = await invokeSnap<SendTransactionResponse>(
-      this.snapId,
-      payload
+  off(event: VegaWalletEvent, callback?: () => void) {
+    this.pollListeners[event] = this.pollListeners[event].filter(
+      (cb) => cb !== callback
     );
+  }
```

Comment:
Suggest extending the `EventEmitter` class rather than hand rolling as it's already bundled?

This is also duplicated across classes, so by extending EventEmitter we can remove this duplication.

Reply:
Addressed in https://github.com/vegaprotocol/frontend-monorepo/pull/5815/commits/e0192f9780f78108cd4d09ee16557ecac93cc512